keps/sig-network/3836-kube-proxy-improved-ingress-connectivity-reliability/README.md

```diff
@@ -0,0 +1,346 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The service controller in the Kubernetes cloud controller manager (KCCM)
+configures the load balancer and its corresponding health check (HC) following
+service related events. The configured health check is then used by the load
+balancer as to determine which instances are candidates for traffic load
+balancing. For certain cloud providers (GCP): the KCCM configures the HC to
+target the Kubernetes service proxy for this information. In this KEP we will
+focus on Kube-proxy, since it's the only service proxy under the responsibility
+of the Kubernetes project.
+
+We can define two classes of services for what concerns the HC:
+
+- A) services which have `externalTrafficPolicy: Cluster`
+- B) services which have `externalTrafficPolicy: Local`
+
+For class A) services: Kube-proxy currently returns an answer following its
+`healthz` state. For class B) services: Kube-proxy only reports if the service
+for which the load balancer was created for, has a `Ready` endpoint running on
+the node.
+
+This KEP proposes two changes:
+
+1. That Kube-proxy's includes the `healthz` state AND its current answer w.r.t
+  the local endpoints, when it answers to the HC for class B) services.
+
+2. That Kube-proxy enables connection draining for terminating Nodes. This is to
+  be done by Kube-proxy by inspecting a field on the Node object which indicates
+  that the "node is terminating/deleting" and when seen, starts failing the HC.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+We need to agree on this field. `deletionTimestamp` is the perfect indicator for
+this, but....the Node object does not have a finalizer, so that wouldn't help.
+@thockin advocates for `.spec.unschedulable`, but I am inclined to say "that
+doesn't indicate if a Node is going away". `unschedulable` just says that no
+workloads can be assigned to the Node anymore. It is true that it's usually
+followed up by a drain...but I am not sure it's good to implement something
+which can easily kill ingress connectivity to a Node / cluster by simply doing
+`kubectl cordon node ...` (It would kill ingress connectivity to the entire
```

Comment:
If the ingress controller is `eTP: Local` it will not go into draining mode (because that will naturally happen when the # of pods -> 0).

If the ingress controller is `eTP: Cluster` the *node* will go into connection-draining, but the pod itself is still Ready, and other nodes can target it (cordon node A; external LB -> node B -> pod on node A).  The connection draining is to gracefully end connections which come in _through_ this node.

`unschedulable` is imperfect, but the worst case is that this node stops acting as an LB backend for a while.  If it becomes schedulable again, running connections are undamaged and new connections can resume.

We could say the same for node-unready (it's not a leading indicator for node going away, but could be an indicator for other trouble -- or not).

Basically, the cost of getting HC wrong for `eTP: Cluster` is relatively low.

Reply:
we have `networkNotReady` , this seems to be Service Network Not Ready

Reply:
> (because that will naturally happen when the # of pods -> 0).

I was talking about only cordoning the node in this case. I.e: not followed by a drain. 

> we have `networkNotReady` , this seems to be Service Network Not Ready

But that's tied to the network plugin, not the service proxy, right?

Reply:
> I was talking about only cordoning the node in this case. I.e: not followed by a drain.

If the ingress controller is eTP: Local it will not go into draining mode - the HC will pass because this node still has endpoints.  If you then drain the node, the HC will fail when pods -> 0.  If you are not draining the node, there's no impact.  So `unschedulable` is fine.

If the ingress controller is eTP: Cluster the node will go into connection-draining, but the pod itself is still Ready, and other nodes can still target it (cordon node A; external LB -> node B -> pod on node A).  If you then drain the node, the pod itself has termination grace period.  If you are not draining the node, there's no impact.  So `unschedulable` is fine.

My point was that your concern "It would kill ingress connectivity to the entire cluster" is untrue.

Reply:
> If the ingress controller is eTP: Local it will not go into draining mode - the HC will pass because this node still has endpoints.

The KEP says above that for eTP:Local the health should be based on both endpoints _and_ kube-proxy health, doesn't it?

Though I can't actually come up with a scenario where that makes more sense than depending just on endpoint readiness/count...

One idea I had is that if kube-proxy has stopped getting updates from the apiserver, then it won't be able to know if the local endpoints are terminating/unready, so it might keep accepting connections, only to have the endpoint pod abruptly die. So in that case it would be good to have the LB stop sending to the pod when kube-proxy reports unhealthy. EXCEPT... kube-proxy doesn't actually report that it's unhealthy if it stops getting updates from the apiserver, so...

Reply:
> The KEP says above that for eTP:Local the health should be based on both endpoints and kube-proxy health, doesn't it?

So, yes: like Dan said. This is what this KEP aims to change. And in that case we fail the HC when the node only is cordoned (unschedulable), causing all new connections to the ingress controller sitting on that Node to fail. And people might choose to put a node in the `unschedulable` state for a variety of reasons, many of which I can't even imagine while I'm writing this. That's why I find it scary.  

I think my company actually does that when we upgrade Kube at one point. We manage the node pools where our workloads run and we do this in manual mode: so when we need to upgrade them we create a new node pool with version N + 1 , then we cordon all existing Nodes in the cluster, but don't evict them, we call some service which will evict them later. But we expect ingress connectivity to work on the unschedulable nodes until that eviction service has kicked in a decided to trigger a restart (which might be minutes / hours)....killing ingress to these workloads for that time, would be bad.

Reply:
I think there's some crossed wires.  We have two different types of health signals:

1) the kube-proxy's own health (currently based on recency of DP updates).
2) whether a given service has andpoints on this node or not.

We (well, some clouds) use the first as LBHC for `eTP:Cluster` services.
Presumably everyone uses the second as LBHC for `eTP:Local`.

Agree so far?

I'm advocating, first and foremost, that we expand `eTP:Cluster` LBHC to include some other "eager" signal (e.g. `unschedulable`) that the node is likely to go away soon.  This has no bearing on the availability of the pods on that node, only on "thru traffic" which uses the node as a 2nd-hop LB.

I'm advocating, though less strenuously, that we expand `eTP:Local` LBHC to include other indication(s) of the node's ability to handle traffic (e.g. DP recency).  I am *NOT* advocating to include `unschedulable` for this.  This could impact the availability of the pods on that node, so we must choose wisely (or choose to ONLY consider endpoints-available).

IOW:

```
func etpClusterHealth() bool {
    return dataplaneHealthy() && !nodeGoingAway()
}

func etpLocalHealth(svc string) bool {
    return haveEndpoints(svc) && dataplaneHealthy()
}
```

`nodeGoingAway()` can be as simple as `return node.spec.unschedulable` but there's no reason not to make it consider more, I think.

* unschedulable is often followed by drain and delete
* ToBeDeleted is usually followed by drain and delete, but is also usually set at the same time as unschedulable
* deletionTimestamp clearly indicates impending death
* The "exclude from LB" label *should* cause the controller to remove the node, but we could also use it here
* ...

Does that help?

Reply:
Yes, that makes sense.